File: electron/providers/tools/write.ts
/**
 * Write Tool - Write content to a file
 */

import * as fs from "fs";
import * as path from "path";

export interface WriteInput {
  file_path: string;
  content: string;
}

export interface WriteOutput {
  success: boolean;
  bytesWritten: number;
}

export function writeTool(input: WriteInput, cwd: string): WriteOutput {
  const filePath = path.isAbsolute(input.file_path)
    ? input.file_path
    : path.resolve(cwd, input.file_path);

  // Ensure parent directory exists
  const dir = path.dirname(filePath);
  if (!fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true });
  }

  fs.writeFileSync(filePath, input.content, "utf-8");

  return {
    success: true,
    bytesWritten: Buffer.byteLength(input.content, "utf-8"),
  };
}

File: src-tauri/src/providers/tools/write.rs
use std::fs;
use std::path::Path;

use anyhow::Result;

/// Write content to a file, creating parent directories as needed.
///
/// * `file_path` - Path to the file (absolute or relative to `cwd`).
/// * `content` - The content to write.
/// * `cwd` - Current working directory for resolving relative paths.
///
/// Returns the number of bytes written.
pub fn write_tool(
    file_path: &str,
    content: &str,
    cwd: &str,
) -> Result<usize> {
    let path = if Path::new(file_path).is_absolute() {
        file_path.to_string()
    } else {
        Path::new(cwd).join(file_path).to_string_lossy().to_string()
    };

    // Ensure parent directory exists
    if let Some(parent) = Path::new(&path).parent() {
        if !parent.exists() {
            fs::create_dir_all(parent)?;
        }
    }

    fs::write(&path, content)?;

    Ok(content.len())
}
